clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -4656,7 +4656,7 @@ void SYCLIntegrationHeader::emit(raw_ostream &O) {
   // whose sole purpose is to run its constructor before the application's
   // main() function.
 
-  if (S.getSyclIntegrationFooter().isDeviceGlobalsEmitted()) {
+  if (NeedToEmitDeviceGlobalRegistration) {
```

Comment:
Can you explain why this change forces emission? Prior to this PR, the variable signaling emit required was maintained in footer and now it is in header. How did changing this solve the issue?

Reply:
Integration header is created when it is accessed via `Sema::getSyclIntegrationHeader()`. Prior to this change this access happened only when info about kernels was communicated to the integration header, so the integration header was not emitted if there was no kernels in user's code. This change forces creation of the header by calling of `Sema::getSyclIntegrationHeader()` by the function that emits integration footer (see line 5108 in SemaSYCL.cpp in this diff). Moving of variable signaling that `__sycl_device_global_registration` is required from footer to header seemed logical choice here, so now the footer communicates some info to the header and therefore forces emission of the header.

Reply:
Thanks for explaining!